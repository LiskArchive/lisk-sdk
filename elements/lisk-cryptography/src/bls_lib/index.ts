/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { BlsLib } from './types';

const blsNotSupportedError = new Error(
	'BLS is not supported on your platform. To enable, please check if optional dependencies are meet.',
);

// eslint-disable-next-line import/no-mutable-exports
export let lib: BlsLib = {
	blsKeyValidate: (_pk: Buffer) => {
		throw blsNotSupportedError;
	},
	blsKeyGen: (_ikm: Buffer) => {
		throw blsNotSupportedError;
	},
	blsSkToPk: (_sk: Buffer) => {
		throw blsNotSupportedError;
	},
	blsAggregate: (_signatures: Buffer[]) => {
		throw blsNotSupportedError;
	},
	blsSign: (_sk: Buffer, _message: Buffer) => {
		throw blsNotSupportedError;
	},
	blsVerify: (_pk: Buffer, _message: Buffer, _signature: Buffer) => {
		throw blsNotSupportedError;
	},
	blsAggregateVerify: (
		_publicKeys: ReadonlyArray<Buffer>,
		_messages: ReadonlyArray<Buffer>,
		_signature: Buffer,
	) => {
		throw blsNotSupportedError;
	},
	blsFastAggregateVerify: (
		_publicKeys: ReadonlyArray<Buffer>,
		_messages: Buffer,
		_signature: Buffer,
	) => {
		throw blsNotSupportedError;
	},
	blsPopProve: (_sk: Buffer) => {
		throw blsNotSupportedError;
	},
	blsPopVerify: (_pk: Buffer, _proof: Buffer) => {
		throw blsNotSupportedError;
	},
};
// eslint-disable-next-line import/no-mutable-exports
export let BLS_SUPPORTED = true;

try {
	// Check if BLS library (optional dependency) available

	// eslint-disable-next-line global-require, import/no-extraneous-dependencies
	require.resolve('@chainsafe/blst');
	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, global-require
	lib = require('./lib');
} catch (err) {
	BLS_SUPPORTED = false;
}

export const {
	blsSign,
	blsVerify,
	blsKeyValidate,
	blsAggregate,
	blsKeyGen,
	blsFastAggregateVerify,
	blsAggregateVerify,
	blsSkToPk,
	blsPopProve,
	blsPopVerify,
} = lib;
