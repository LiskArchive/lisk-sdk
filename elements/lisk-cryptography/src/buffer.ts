/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import * as BigNum from '@liskhq/bignum';

export const bigNumberToBuffer = (bignumber: string, size: number) =>
	new BigNum(bignumber).toBuffer({ size, endian: 'big' });

export const bufferToBigNumberString = (bigNumberBuffer: Buffer): string =>
	BigNum.fromBuffer(bigNumberBuffer).toString();

export const bufferToHex = (buffer: Buffer): string =>
	Buffer.from(buffer).toString('hex');

const hexRegex = /^[0-9a-f]+/i;
export const hexToBuffer = (hex: string, argumentName = 'Argument'): Buffer => {
	if (typeof hex !== 'string') {
		throw new TypeError(`${argumentName} must be a string.`);
	}
	const matchedHex = (hex.match(hexRegex) || [])[0];
	if (!matchedHex || matchedHex.length !== hex.length) {
		throw new TypeError(`${argumentName} must be a valid hex string.`);
	}
	// tslint:disable-next-line no-magic-numbers
	if (matchedHex.length % 2 !== 0) {
		throw new TypeError(
			`${argumentName} must have a valid length of hex string.`,
		);
	}

	return Buffer.from(matchedHex, 'hex');
};
