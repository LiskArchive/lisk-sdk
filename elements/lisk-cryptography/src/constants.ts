/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
export const SIGNED_MESSAGE_PREFIX = 'Lisk Signed Message:\n';
export const BINARY_ADDRESS_LENGTH = 20;
export const DEFAULT_LISK32_ADDRESS_PREFIX = 'lsk';
export const DEFAULT_BASE32_ADDRESS_PREFIX = DEFAULT_LISK32_ADDRESS_PREFIX;
export const ED25519_CURVE = 'ed25519 seed';
export const MAX_UINT32 = 4294967295;
export const HARDENED_OFFSET = 0x80000000;
export const HASH_LENGTH = 32;
