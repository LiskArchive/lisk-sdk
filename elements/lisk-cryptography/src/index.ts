/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import * as constants from './constants';

export * from './buffer';
export * from './convert';
export * from './encrypt';
export * from './hash';
export * from './keys';
export * from './sign';
export * from './hash_onion';
export { getRandomBytes } from './nacl';

export { constants };
