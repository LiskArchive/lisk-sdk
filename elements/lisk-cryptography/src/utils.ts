/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import * as crypto from 'crypto';
import { ED25519_CURVE, MAX_UINT32 } from './constants';

export const readBit = (buf: Buffer, bit: number): boolean => {
	const byteIndex = Math.floor(bit / 8);
	const bitIndex = bit % 8;

	// eslint-disable-next-line no-bitwise
	return (buf[byteIndex] >> bitIndex) % 2 === 1;
};

export const writeBit = (buf: Buffer, bit: number, val: boolean): void => {
	const byteIndex = Math.floor(bit / 8);
	const bitIndex = bit % 8;

	if (val) {
		// eslint-disable-next-line no-bitwise, no-param-reassign
		buf[byteIndex] |= 1 << bitIndex;
	} else {
		// eslint-disable-next-line no-bitwise, no-param-reassign
		buf[byteIndex] &= ~(1 << bitIndex);
	}
};

export const isValidPath = (path: string) => {
	if (!path.startsWith('m') || !path.includes('/')) {
		return false;
	}

	return path
		.split('/')
		.slice(1)
		.every(segment => {
			if (!/^[0-9']+$/g.test(segment)) {
				return false;
			}

			return segment.includes(`'`)
				? parseInt(segment.slice(0, -1), 10) <= MAX_UINT32 / 2
				: parseInt(segment, 10) <= MAX_UINT32 / 2;
		});
};

export const getMasterKeyFromSeed = (seed: Buffer) => {
	const hmac = crypto.createHmac('sha512', ED25519_CURVE);
	const I = hmac.update(seed).digest();
	const IL = I.slice(0, 32);
	const IR = I.slice(32);
	return {
		key: IL,
		chainCode: IR,
	};
};

export const getChildKey = (node: { key: Buffer; chainCode: Buffer }, index: number) => {
	const indexBuffer = Buffer.allocUnsafe(4);
	indexBuffer.writeUInt32BE(index, 0);
	const data = Buffer.concat([Buffer.alloc(1, 0), node.key, indexBuffer]);
	const I = crypto.createHmac('sha512', node.chainCode).update(data).digest();
	const IL = I.slice(0, 32);
	const IR = I.slice(32);

	return {
		key: IL,
		chainCode: IR,
	};
};
