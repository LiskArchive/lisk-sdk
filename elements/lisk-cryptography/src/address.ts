/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { BINARY_ADDRESS_LENGTH, DEFAULT_LISK32_ADDRESS_PREFIX } from './constants';
import { getKeys } from './ed';
import { getPublicKey } from './nacl';
import { hash } from './utils';

const CHARSET = 'zxvcpmbn3465o978uyrtkqew2adsjhfg';

const convertUIntArray = (uintArray: number[], fromBits: number, toBits: number): number[] => {
	// eslint-disable-next-line no-bitwise
	const maxValue = (1 << toBits) - 1;
	let accumulator = 0;
	let bits = 0;
	const result = [];
	// eslint-disable-next-line
	for (let p = 0; p < uintArray.length; p += 1) {
		const byte = uintArray[p];
		// check that the entry is a value between 0 and 2^frombits-1
		// eslint-disable-next-line no-bitwise
		if (byte < 0 || byte >> fromBits !== 0) {
			return [];
		}

		// eslint-disable-next-line no-bitwise
		accumulator = (accumulator << fromBits) | byte;
		bits += fromBits;
		while (bits >= toBits) {
			bits -= toBits;
			// eslint-disable-next-line no-bitwise
			result.push((accumulator >> bits) & maxValue);
		}
	}

	return result;
};

const convertUInt5ToBase32 = (uint5Array: number[]): string =>
	uint5Array.map((val: number) => CHARSET[val]).join('');

export const getAddressFromPublicKey = (publicKey: Buffer): Buffer => {
	const buffer = hash(publicKey);
	const truncatedBuffer = buffer.slice(0, BINARY_ADDRESS_LENGTH);

	if (truncatedBuffer.length !== BINARY_ADDRESS_LENGTH) {
		throw new Error('The Lisk addresses must contains exactly 20 bytes');
	}

	return truncatedBuffer;
};

export const getAddressAndPublicKeyFromPassphrase = (
	passphrase: string,
): { readonly address: Buffer; readonly publicKey: Buffer } => {
	const { publicKey } = getKeys(passphrase);
	const address = getAddressFromPublicKey(publicKey);

	return {
		address,
		publicKey,
	};
};

export const getAddressFromPassphrase = (passphrase: string): Buffer => {
	const { publicKey } = getKeys(passphrase);

	return getAddressFromPublicKey(publicKey);
};

export const getAddressFromPrivateKey = (privateKey: Buffer): Buffer => {
	const publicKey = getPublicKey(privateKey);

	return getAddressFromPublicKey(publicKey);
};

const GENERATOR = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];

// See for details: https://github.com/LiskHQ/lips/blob/master/proposals/lip-0018.md#creating-checksum
const polymod = (uint5Array: number[]): number => {
	let chk = 1;
	for (const value of uint5Array) {
		// eslint-disable-next-line no-bitwise
		const top = chk >> 25;
		// eslint-disable-next-line no-bitwise
		chk = ((chk & 0x1ffffff) << 5) ^ value;
		for (let i = 0; i < 5; i += 1) {
			// eslint-disable-next-line no-bitwise
			if ((top >> i) & 1) {
				// eslint-disable-next-line no-bitwise
				chk ^= GENERATOR[i];
			}
		}
	}

	return chk;
};

const createChecksum = (uint5Array: number[]): number[] => {
	const values = uint5Array.concat([0, 0, 0, 0, 0, 0]);
	// eslint-disable-next-line no-bitwise
	const mod = polymod(values) ^ 1;
	const result = [];
	for (let p = 0; p < 6; p += 1) {
		// eslint-disable-next-line no-bitwise
		result.push((mod >> (5 * (5 - p))) & 31);
	}
	return result;
};

const verifyChecksum = (integerSequence: number[]): boolean => polymod(integerSequence) === 1;

const addressToLisk32 = (address: Buffer): string => {
	const byteSequence = [];
	for (const b of address) {
		byteSequence.push(b);
	}
	const uint5Address = convertUIntArray(byteSequence, 8, 5);
	const uint5Checksum = createChecksum(uint5Address);
	return convertUInt5ToBase32(uint5Address.concat(uint5Checksum));
};

export const getLisk32AddressFromPublicKey = (
	publicKey: Buffer,
	prefix = DEFAULT_LISK32_ADDRESS_PREFIX,
): string => `${prefix}${addressToLisk32(getAddressFromPublicKey(publicKey))}`;

export const getLisk32AddressFromPassphrase = (
	passphrase: string,
	prefix = DEFAULT_LISK32_ADDRESS_PREFIX,
): string => {
	const { publicKey } = getAddressAndPublicKeyFromPassphrase(passphrase);
	return getLisk32AddressFromPublicKey(publicKey, prefix);
};

const LISK32_ADDRESS_LENGTH = 41;
const LISK32_CHARSET = 'zxvcpmbn3465o978uyrtkqew2adsjhfg';

export const validateLisk32Address = (
	address: string,
	prefix = DEFAULT_LISK32_ADDRESS_PREFIX,
): true | never => {
	if (address.length !== LISK32_ADDRESS_LENGTH) {
		throw new Error('Address length does not match requirements. Expected 41 characters.');
	}

	const addressPrefix = address.substring(0, 3);

	if (addressPrefix !== prefix) {
		throw new Error(
			`Invalid address prefix. Actual prefix: ${addressPrefix}, Expected prefix: ${prefix}`,
		);
	}

	const addressSubstringArray = address.substring(3).split('');

	if (!addressSubstringArray.every(char => LISK32_CHARSET.includes(char))) {
		throw new Error(
			"Invalid character found in address. Only allow characters: 'abcdefghjkmnopqrstuvwxyz23456789'.",
		);
	}

	const integerSequence = addressSubstringArray.map(char => LISK32_CHARSET.indexOf(char));

	if (!verifyChecksum(integerSequence)) {
		throw new Error('Invalid checksum for address.');
	}

	return true;
};

export const getAddressFromLisk32Address = (
	base32Address: string,
	prefix = DEFAULT_LISK32_ADDRESS_PREFIX,
): Buffer => {
	validateLisk32Address(base32Address, prefix);
	// Ignore lsk prefix and checksum
	const base32AddressNoPrefixNoChecksum = base32Address.substring(
		prefix.length,
		base32Address.length - 6,
	);

	const addressArray = base32AddressNoPrefixNoChecksum.split('');
	const integerSequence = addressArray.map(char => LISK32_CHARSET.indexOf(char));
	const integerSequence8 = convertUIntArray(integerSequence, 5, 8);

	return Buffer.from(integerSequence8);
};

export const getLisk32AddressFromAddress = (
	address: Buffer,
	prefix = DEFAULT_LISK32_ADDRESS_PREFIX,
): string => `${prefix}${addressToLisk32(address)}`;
