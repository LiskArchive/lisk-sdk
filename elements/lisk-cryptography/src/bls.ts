/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import * as crypto from 'crypto';
import { Mnemonic } from '@liskhq/lisk-passphrase';
import {
	blsSign,
	blsVerify,
	blsKeyValidate,
	blsAggregate,
	blsKeyGen,
	blsFastAggregateVerify,
	blsSkToPk,
	BLS_SUPPORTED,
	blsPopVerify,
	blsPopProve,
} from './bls_lib';
import { hash, parseKeyDerivationPath, tagMessage } from './utils';
import { EMPTY_SALT, HASH_LENGTH, L, SHA256 } from './constants';

export { BLS_SUPPORTED };
export const generatePrivateKey = blsKeyGen;
export const getPublicKeyFromPrivateKey = blsSkToPk;
export const validateKey = blsKeyValidate;
export const popVerify = blsPopVerify;
export const popProve = blsPopProve;

const readBit = (buf: Buffer, bit: number): boolean => {
	const byteIndex = Math.floor(bit / 8);
	const bitIndex = bit % 8;

	// eslint-disable-next-line no-bitwise
	return (buf[byteIndex] >> bitIndex) % 2 === 1;
};

const writeBit = (buf: Buffer, bit: number, val: boolean): void => {
	const byteIndex = Math.floor(bit / 8);
	const bitIndex = bit % 8;

	if (val) {
		// eslint-disable-next-line no-bitwise, no-param-reassign
		buf[byteIndex] |= 1 << bitIndex;
	} else {
		// eslint-disable-next-line no-bitwise, no-param-reassign
		buf[byteIndex] &= ~(1 << bitIndex);
	}
};

export const signData = (tag: string, chainID: Buffer, data: Buffer, privateKey: Buffer): Buffer =>
	blsSign(privateKey, hash(tagMessage(tag, chainID, data)));

export const verifyData = (
	tag: string,
	chainID: Buffer,
	data: Buffer,
	signature: Buffer,
	publicKey: Buffer,
): boolean => blsVerify(publicKey, hash(tagMessage(tag, chainID, data)), signature);

export const createAggSig = (
	publicKeysList: Buffer[],
	pubKeySignaturePairs: { publicKey: Buffer; signature: Buffer }[],
): { aggregationBits: Buffer; signature: Buffer } => {
	const aggregationBits = Buffer.alloc(Math.ceil(publicKeysList.length / 8));
	const signatures: Buffer[] = [];

	for (const pair of pubKeySignaturePairs) {
		signatures.push(pair.signature);
		const index = publicKeysList.findIndex(key => key.equals(pair.publicKey));
		writeBit(aggregationBits, index, true);
	}
	const signature = blsAggregate(signatures);

	if (!signature) {
		throw new Error('Can not aggregate signatures');
	}

	return { aggregationBits, signature };
};

export const verifyAggSig = (
	publicKeysList: Buffer[],
	aggregationBits: Buffer,
	signature: Buffer,
	tag: string,
	chainID: Buffer,
	message: Buffer,
): boolean => {
	const taggedMessage = tagMessage(tag, chainID, message);
	const keys: Buffer[] = [];

	for (const [index, key] of publicKeysList.entries()) {
		if (readBit(aggregationBits, index)) {
			keys.push(key);
		}
	}

	return blsFastAggregateVerify(keys, hash(taggedMessage), signature);
};

export const verifyWeightedAggSig = (
	publicKeysList: Buffer[],
	aggregationBits: Buffer,
	signature: Buffer,
	tag: string,
	chainID: Buffer,
	message: Buffer,
	weights: number[] | bigint[],
	threshold: number | bigint,
): boolean => {
	const taggedMessage = tagMessage(tag, chainID, message);
	const keys: Buffer[] = [];
	let weightSum = BigInt(0);

	for (const [index, key] of publicKeysList.entries()) {
		if (readBit(aggregationBits, index)) {
			keys.push(key);
			weightSum += BigInt(weights[index]);
		}
	}

	if (weightSum < BigInt(threshold)) {
		return false;
	}

	return blsFastAggregateVerify(keys, hash(taggedMessage), signature);
};

// eslint-disable-next-line no-bitwise
const flipBits = (buf: Buffer) => Buffer.from(buf.map(x => x ^ 0xff));

const sha256 = (x: Buffer) => crypto.createHash(SHA256).update(x).digest();

const hmacSHA256 = (key: Buffer, message: Buffer, hashValue: string) =>
	crypto.createHmac(hashValue, key).update(message).digest();

const hkdfSHA256 = (ikm: Buffer, length: number, salt: Buffer, info: Buffer) => {
	if (salt.length === 0) {
		// eslint-disable-next-line no-param-reassign
		salt = EMPTY_SALT;
	}
	const PRK = hmacSHA256(salt, ikm, SHA256);
	let t = Buffer.from([]);
	let OKM = Buffer.from([]);

	for (let i = 0; i < Math.ceil(length / HASH_LENGTH); i += 1) {
		t = hmacSHA256(PRK, Buffer.concat([t, info, Buffer.from([1 + i])]), SHA256);
		OKM = Buffer.concat([OKM, t]);
	}
	return OKM.slice(0, length);
};

const toLamportSK = (IKM: Buffer, salt: Buffer) => {
	const info = Buffer.from([]);
	const OKM = hkdfSHA256(IKM, L, salt, info);

	const lamportSK = [];
	for (let i = 0; i < 255; i += 1) {
		lamportSK.push(OKM.slice(i * 32, (i + 1) * 32));
	}
	return lamportSK;
};

const parentSKToLamportPK = (parentSK: Buffer, index: number) => {
	const salt = Buffer.allocUnsafe(4);
	salt.writeUIntBE(index, 0, 4);

	const IKM = parentSK;
	const hashedLamport0 = toLamportSK(IKM, salt).map(x => sha256(x));
	const hashedLamport1 = toLamportSK(flipBits(IKM), salt).map(x => sha256(x));

	const lamportPK = Buffer.concat(hashedLamport0.concat(hashedLamport1));
	return sha256(lamportPK);
};

const deriveChildSK = (parentSK: Buffer, index: number) => {
	const lamportPK = parentSKToLamportPK(parentSK, index);
	return blsKeyGen(lamportPK);
};

export const getPrivateKeyFromPhraseAndPath = async (phrase: string, path: string) => {
	const masterSeed = await Mnemonic.mnemonicToSeed(phrase);
	let key = blsKeyGen(masterSeed);

	for (const segment of parseKeyDerivationPath(path)) {
		key = deriveChildSK(key, segment);
	}

	return key;
};
