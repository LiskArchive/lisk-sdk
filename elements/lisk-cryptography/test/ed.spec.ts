/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { makeInvalid } from './helpers';
import {
	signMessageWithPrivateKey,
	verifyMessageWithPublicKey,
	printSignedMessage,
	signAndPrintMessage,
	signDataWithPrivateKey,
	verifyData,
	getPrivateKeyFromPhraseAndPath,
	getPublicKeyFromPrivateKey,
} from '../src/ed';
import { createMessageTag } from '../src/utils';
import { MAX_UINT32 } from '../src/constants';

const changeLength = (buffer: Buffer): Buffer => Buffer.concat([Buffer.from('00', 'hex'), buffer]);

const privateKey = Buffer.from(
	'314852d7afb0d4c283692fef8a2cb40e30c7a5df2ed79994178c10ac168d6d977ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588',
	'hex',
);
const publicKey = Buffer.from(
	'7ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588',
	'hex',
);

const tag = createMessageTag('TST');

describe('sign and verify message', () => {
	const message = 'Some default text.';
	const messageBuffer = Buffer.from(message, 'utf8');
	const { signature } = signMessageWithPrivateKey(message, privateKey);
	const { signature: signatureCustomTag } = signMessageWithPrivateKey(message, privateKey, tag);
	const printedMessage = `
-----BEGIN LISK SIGNED MESSAGE-----
-----MESSAGE-----
${message}
-----PUBLIC KEY-----
${publicKey.toString('hex')}
-----SIGNATURE-----
${signature.toString('hex')}
-----END LISK SIGNED MESSAGE-----
`.trim();

	describe('#signMessageWithPrivateKey', () => {
		it('should create a signed message using a private key', () => {
			const signedStringMessage = signMessageWithPrivateKey(message, privateKey);
			expect(signedStringMessage).toMatchSnapshot();

			const signedBufferMessage = signMessageWithPrivateKey(messageBuffer, privateKey);
			expect(signedBufferMessage.signature).toEqual(signedStringMessage.signature);
		});

		it('should create a signed message using a private key and a custom tag', () => {
			const signedMessage = signMessageWithPrivateKey(message, privateKey, tag);
			expect(signedMessage).toMatchSnapshot();
		});
	});

	describe('#verifyMessageWithPublicKey', () => {
		it('should detect invalid publicKeys', () => {
			expect(() =>
				verifyMessageWithPublicKey({ message, signature, publicKey: changeLength(publicKey) }),
			).toThrow('Invalid publicKey, expected 32-byte publicKey');
		});

		it('should detect invalid signatures', () => {
			expect(() =>
				verifyMessageWithPublicKey({ message, signature: changeLength(signature), publicKey }),
			).toThrow('Invalid signature length, expected 64-byte signature');
		});

		it('should return false if the signature is invalid', () => {
			const verification = verifyMessageWithPublicKey({
				message,
				signature: makeInvalid(signature),
				publicKey,
			});
			expect(verification).toBeFalse();
		});

		it('should return true if the signature is valid and message is in string format', () => {
			const signedMessage = { message, publicKey, signature };
			expect(verifyMessageWithPublicKey(signedMessage)).toBeTrue();
		});

		it('should return true if the signature is valid and message is in Buffer format', () => {
			const signedMessage = { message: messageBuffer, publicKey, signature };
			expect(verifyMessageWithPublicKey(signedMessage)).toBeTrue();
		});

		it('should return true if the signature is valid and a custom tag is used', () => {
			const signedMessage = { message, publicKey, signature: signatureCustomTag, tag };
			expect(verifyMessageWithPublicKey(signedMessage)).toBeTrue();
		});
	});

	describe('#printSignedMessage', () => {
		it('should wrap a single signed message into a printed Lisk template', () => {
			const generatedPrintedMessage = printSignedMessage({ message, signature, publicKey });
			expect(generatedPrintedMessage).toBe(printedMessage);
		});
	});

	describe('#signAndPrintMessage', () => {
		it('should sign the message once and wrap it into a printed Lisk template', () => {
			const signedAndPrintedMessage = signAndPrintMessage(message, privateKey);
			expect(signedAndPrintedMessage).toBe(printedMessage);
		});
	});
});

describe('sign and verify data', () => {
	const chainID = Buffer.from('10000000', 'hex');

	const data = Buffer.from('This is some data');
	const signature = signDataWithPrivateKey(tag, chainID, data, privateKey);

	describe('#signDataWithPrivateKey', () => {
		it('should sign data', () => {
			expect(signDataWithPrivateKey(tag, chainID, data, privateKey)).toMatchSnapshot();
		});
	});

	describe('#verifyData', () => {
		it('should return false for an invalid signature', () => {
			const verification = verifyData(tag, chainID, data, makeInvalid(signature), publicKey);
			expect(verification).toBeFalse();
		});

		it('should return true for a valid signature', () => {
			const verification = verifyData(tag, chainID, data, signature, publicKey);
			expect(verification).toBeTrue();
		});
	});
});

describe('getPrivateKeyFromPhraseAndPath', () => {
	const passphrase =
		'target cancel solution recipe vague faint bomb convince pink vendor fresh patrol';

	it('should get keypair from valid phrase and path', async () => {
		const privateKeyFromPassphrase = await getPrivateKeyFromPhraseAndPath(
			passphrase,
			`m/44'/134'/0'`,
		);
		const publicKeyFromPrivateKey = getPublicKeyFromPrivateKey(privateKeyFromPassphrase);
		expect(publicKeyFromPrivateKey.toString('hex')).toBe(
			'c6bae83af23540096ac58d5121b00f33be6f02f05df785766725acdd5d48be9d',
		);
		expect(privateKeyFromPassphrase.toString('hex')).toBe(
			'c465dfb15018d3aef0d94d411df048e240e87a3ec9cd6d422cea903bfc101f61c6bae83af23540096ac58d5121b00f33be6f02f05df785766725acdd5d48be9d',
		);
	});

	it('should derive distinct keys from same valid phrase but distinct paths', async () => {
		const privateKeyFromPassphrase = await getPrivateKeyFromPhraseAndPath(
			passphrase,
			`m/44'/134'/0'`,
		);

		const anotherPrivateKeyFromPassphrase = await getPrivateKeyFromPhraseAndPath(
			passphrase,
			`m/44'/134'/1'`,
		);

		expect(privateKeyFromPassphrase).not.toEqual(anotherPrivateKeyFromPassphrase);
	});

	it('should fail for empty string path', async () => {
		await expect(getPrivateKeyFromPhraseAndPath(passphrase, '')).rejects.toThrow(
			'Invalid path format',
		);
	});

	it('should fail if path does not start with "m"', async () => {
		await expect(getPrivateKeyFromPhraseAndPath(passphrase, `/44'/134'/0'`)).rejects.toThrow(
			'Invalid path format',
		);
	});

	it('should fail if path does not include at least one "/"', async () => {
		await expect(getPrivateKeyFromPhraseAndPath(passphrase, 'm441340')).rejects.toThrow(
			'Invalid path format',
		);
	});

	it('should fail for path with invalid segment', async () => {
		await expect(
			getPrivateKeyFromPhraseAndPath(
				passphrase,
				`m//134'/0'`, // should be number with or without ' between every back slash
			),
		).rejects.toThrow('Invalid path format');
	});

	it('should fail for path with invalid characters', async () => {
		await expect(getPrivateKeyFromPhraseAndPath(passphrase, `m/a'/134b'/0'`)).rejects.toThrow(
			'Invalid path format',
		);
	});

	it('should fail for path with non-sanctioned special characters', async () => {
		await expect(getPrivateKeyFromPhraseAndPath(passphrase, `m/4a'/#134b'/0'`)).rejects.toThrow(
			'Invalid path format',
		);
	});

	it(`should fail for path with segment greater than ${MAX_UINT32} / 2`, async () => {
		await expect(
			getPrivateKeyFromPhraseAndPath(passphrase, `m/44'/134'/${MAX_UINT32}'`),
		).rejects.toThrow('Invalid path format');
	});
});
