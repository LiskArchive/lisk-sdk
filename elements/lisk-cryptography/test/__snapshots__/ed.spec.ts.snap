// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sign and verify data #signDataWithPrivateKey should sign a transaction 1`] = `
{
  "data": [
    44,
    123,
    78,
    103,
    191,
    106,
    112,
    48,
    222,
    154,
    116,
    84,
    229,
    225,
    107,
    141,
    35,
    220,
    238,
    31,
    94,
    164,
    184,
    184,
    129,
    200,
    158,
    197,
    131,
    77,
    83,
    74,
    254,
    61,
    121,
    26,
    178,
    228,
    93,
    243,
    49,
    214,
    68,
    147,
    83,
    124,
    229,
    116,
    72,
    2,
    93,
    9,
    207,
    219,
    163,
    155,
    167,
    237,
    238,
    213,
    5,
    239,
    182,
    2,
  ],
  "type": "Buffer",
}
`;

exports[`sign and verify message #signMessageWithPrivateKey should create a signed message using a private key 1`] = `
{
  "message": "Some default text.",
  "publicKey": {
    "data": [
      126,
      244,
      92,
      213,
      37,
      233,
      91,
      122,
      134,
      36,
      75,
      189,
      78,
      180,
      85,
      9,
      20,
      173,
      6,
      48,
      16,
      19,
      149,
      143,
      77,
      214,
      77,
      50,
      239,
      123,
      197,
      136,
    ],
    "type": "Buffer",
  },
  "signature": {
    "data": [
      216,
      59,
      84,
      93,
      38,
      134,
      126,
      198,
      234,
      63,
      68,
      165,
      12,
      18,
      97,
      95,
      129,
      57,
      231,
      69,
      112,
      60,
      207,
      130,
      153,
      74,
      208,
      30,
      37,
      130,
      116,
      119,
      208,
      64,
      81,
      129,
      54,
      18,
      212,
      230,
      116,
      75,
      145,
      156,
      163,
      189,
      218,
      124,
      21,
      106,
      33,
      184,
      240,
      156,
      154,
      151,
      83,
      3,
      75,
      88,
      126,
      163,
      43,
      7,
    ],
    "type": "Buffer",
  },
}
`;
